qadence/engines/jax/differentiable_backend.py

```diff
@@ -0,0 +1,151 @@
+        endianness: Endianness = Endianness.BIG,
+    ) -> ArrayLike:
+        """Run on the underlying backend."""
+        return self.backend.run(
+            circuit=circuit, param_values=param_values, state=state, endianness=endianness
+        )
+
+    def expectation(
+        self,
+        circuit: ConvertedCircuit,
+        observable: list[ConvertedObservable] | ConvertedObservable,
+        param_values: ParamDictType = {},
+        state: Array | None = None,
+        measurement: Measurements | None = None,
+        noise: Noise | None = None,
+        mitigation: Mitigations | None = None,
+        endianness: Endianness = Endianness.BIG,
+    ) -> ArrayLike:
+        observable = observable if isinstance(observable, list) else [observable]
+
+        if self.diff_mode == DiffMode.AD:
+            expectation = self.backend.expectation(circuit, observable, param_values, state)
+        else:
+            expectation = JaxDifferentiableExpectation(
+                backend=self.backend,
+                circuit=circuit,
+                observable=observable,
+                param_values=param_values,
+                state=state,
+                measurement=measurement,
+                noise=noise,
+                mitigation=mitigation,
+                endianness=endianness,
+            ).psr()
+        return expectation
+
+    def sample(
+        self,
+        circuit: ConvertedCircuit,
+        param_values: ParamDictType = {},
+        n_shots: int = 100,
+        state: ArrayLike | None = None,
+        noise: Noise | None = None,
+        mitigation: Mitigations | None = None,
+        endianness: Endianness = Endianness.BIG,
+    ) -> list[Counter]:
+        """Sample bitstring from the registered circuit.
+
+        Arguments:
+            circuit: A backend native quantum circuit to be executed.
+            param_values: The values of the parameters after embedding
+            n_shots: The number of shots. Defaults to 1.
+            state: Initial state.
+            noise: A noise model to use.
+            mitigation: A mitigation protocol to apply to noisy samples.
+            endianness: Endianness of the resulting bitstrings.
+
+        Returns:
+            An iterable with all the sampled bitstrings
+        """
+
+        return self.backend.sample(
+            circuit=circuit,
+            param_values=param_values,
+            n_shots=n_shots,
+            state=state,
+            noise=noise,
+            mitigation=mitigation,
+            endianness=endianness,
+        )
+
+    def circuit(self, circuit: QuantumCircuit) -> ConvertedCircuit:
+        parametrized_blocks = list(uuid_to_block(circuit.block).values())
+        non_prim_blocks = filter(lambda b: not isinstance(b, PrimitiveBlock), parametrized_blocks)
+        if len(list(non_prim_blocks)) > 0:
+            raise ValueError(
+                "The circuit contains non-primitive blocks that are currently not supported by the "
```

Comment:
Out of curiosity, what would be an example of a non-primitive parametrised block ? Multi-parametrised ?

Reply:
good question